pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+                for i, t in enumerate(tapes_min):
+                    with t:
+                        qml.PauliRot(
+                            -np.pi / 2,
+                            self.lie_algebra_basis_names[i],
+                            wires=list(range(self.nqubits)),
+                        )
+                        for o in obs:
+                            qml.expval(o)
+                tapes.append((tapes_plus, tapes_min))
+            return tapes, None
+
+        self.algebra_commutator = algebra_commutator
+
+    def step_and_cost(
+        self,
+        *args,
+        **kwargs,
+    ):
+        r"""Update the circuit with one step of the optimizer and return the corresponding
+        objective function value prior to the step.
+
+        Args:
+            *args:
+            **kwargs:
+
+        Returns:
+
+        """
+        omegas = self.get_omegas()
+        non_zero_lie_algebra_elements = []
+        non_zero_omegas = []
+        for i, element in enumerate(omegas):
+            if not np.isclose(element, 0):
+                non_zero_lie_algebra_elements.append(
+                    tuple(
+                        self.pauli_str_to_int[p]
+                        for p in self.lie_algebra_basis_names[i]
+                    )
+                )
+                non_zero_omegas.append(-omegas[i])
+        lie_gradient = qml.Hamiltonian(
+            non_zero_omegas,
+            [self.get_pauli_op(ps) for ps in non_zero_lie_algebra_elements],
+        )
+        temp_circuit = copy.copy(self.circuit)
+
+        @append_time_evolution(lie_gradient, self.stepsize)
+        def circuit(params, wires, **kwargs):
+            """Append the time evolution"""
+            for obj in temp_circuit.qtape.operations:
+                qml.apply(obj)
+
+            return qml.state()
+
+        # TODO: qml.map is not working here, it throws an error for pauli @ pauli objects, which have `name` tensor.
+        # cost_fn = qml.map(circuit, self.observables, device=self.circuit.device, measure='expval')([],[])
+        # cost_fn = qml.QNodeCollection(circuit, self.observables, device=self.circuit.device, measure='expval')([],[])
+        # cost = sum([self.coeffs[i] * obs for i, obs in enumerate(cost_fn)])
+        # print(cost)
+
+        self.circuit = qml.QNode(circuit, temp_circuit.device)
+        return self.circuit([], [])
+
+    def step(
+        self,
+        *args,
+        **kwargs,
+    ):
+        r"""Update the circuit with one step of the optimizer
+
+        Args:
+            *args:
+            **kwargs:
+
+        Returns:
+
+        """
+        self.step_and_cost(
+            *args,
+            **kwargs,
+        )
+
+        return None
+
+    def get_su_n_operators(self):
+        r"""Get the 2x2 SU(N) operators. The dimension of the group is N^2-1.
+
+        Returns:
+            List of (N^2)x(N^2) numpy complex arrays and corresponding paulis words
+        """
```

Comment:
@therooler I'm curious if some of these Pauli utility functions could be replaced by calling existing functionality in the [`qml.grouping`](https://pennylane.readthedocs.io/en/latest/code/qml_grouping.html) module 🙂 

For example, I _think_ this could be replaced with a call to [`pauli_group`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.grouping.pauli_group.html)?

Reply:
Amazing, yeah I think that would work.